### PR TITLE
[#12] Protect endpoints with authorization

### PR DESCRIPTION
**Pull request by jinju2828**
### What (if any) features are you implementing?
- Protect endpoints with authorization
### What (if anything) did you refactor?
- N/A
### Were there any issues that arose?
- N/A

### Refactor checklist
- [V] Check variable and parameter names: are they descriptive and unique?
- [V] Is the code DRY? SRP?
- [V] Is it easy for others to read your code? (explanatory comments needed?)
- [V] Is the syntax consistent? (e.g. use of semicolons, blank lines)
- [V] Are commit messages formatted correctly?
No Authorization
When Authorization is not sent in header, show "Access denied" error message
Authorized
When Authorization is valid, show welcome message
token
Login test with simple HTML page, token is successfully stored in local storage
login_fail
When login failed, token is not stored.

### CODE REVIEW

**Review by AnaBennett11 — APPROVED** [review]
Thank you for doing this

File: backend/server.js [hunk 1]
```diff
@@ -1,115 +1,121 @@
...
-    });
-  }
 });
 
 //log in
-app.post("/auth/signin", async (req, res) => {
+app.post("/auth/login", async (req, res) => {
     const { email, password } = req.body;
-
+    console.log("email, pw", email, password);
```
Comment (YashsviG) [inline backend/server.js:118]: we shouldnt be logging sensitive data

File: backend/server.js [hunk 2]
```diff
@@ -135,12 +141,19 @@ app.post("/auth/signin", async (req, res) => {
                 { userId: user.userId, userEmail: user.userEmail },
                 secretKey
             );
-            res.status(200).json({ message: 'Login successful', token: token });
+            console.log("logged in");
+            return res.status(200)
+                .header('Authorization', `Bearer ${token}`)
+                .json({ message: 'Login successful', token: token });
```
Comment (YashsviG) [inline backend/server.js:147]: Nice

File: package.json [hunk 3]
```diff
@@ -18,7 +18,9 @@
   },
   "homepage": "https://github.com/cherryontech/squad-debugging-darlings#readme",
   "dependencies": {
+    "axios": "^1.3.4",
     "bcryptjs": "^2.4.3",
+    "cors": "^2.8.5",
     "dotenv": "^16.0.3",
```
Comment (YashsviG) [inline package.json:24]: We shouldn't be having any package.json files outside of the backend/frontend folders. Delete this and the lock file and have your depencies installed within back-end itself

File: backend/auth.js [hunk 4]
```diff
@@ -0,0 +1,33 @@
+const jwt = require('jsonwebtoken');
+
+// define a middleware function to check authentication
+module.exports = (req, res, next) => {
+    // console.log("req only", req);
+    // console.log("req auth headers", req.headers['authorization']);
+    const authHeader = req.headers['authorization'];
+    console.log("authHeader", authHeader);
```
Comment (YashsviG) [inline backend/auth.js:8]: delete this log, we dont need any sensitive data as such in prod

File: backend/auth.js [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    jwt.verify(token, process.env.JWT_SECRET, (err, user) => {
+        if (err) {
+            // if the token is not valid, return a 403 Forbidden error
+            return res.sendStatus(403).send("Invalid token");
+            // return next({ status: 403, message: 'Invalid token' });
+        }
+
+        // if the token is valid, attach the user information to the request object and call next middleware
+        req.user = user;
+        next();
```
Comment (YashsviG) [inline backend/auth.js:31]: Nice work!

**Review by YashsviG — CHANGES_REQUESTED** [review]

File: backend/server.js (hunk 1 above)
Comment (jinju2828) [reply backend/server.js:118]: Thanks, will remove it

File: backend/auth.js (hunk 4 above)
Comment (jinju2828) [reply backend/auth.js:8]: Thanks, will remove it

**Review by YashsviG — APPROVED** [review]
Love the work!

**Pull request merged**